### PR TITLE
Add an inbuilt facility to perform write and fetch retries

### PR DESCRIPTION
**Pull request by robskillington**
This change adds the ability to set a retrier for writes and fetches.  It allows for anyone to use an out of the box retrier from the `github.com/m3db/m3x/retry` package.  It also let's them create their own custom retrier that fits the `Retrier` interface and specify that instead.

cc @xichen2020 @cw9 @prateek @ben-lerner

### CODE REVIEW

File: client/fetch_attempt.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+func (p *fetchAttemptPool) Init() {
+	p.pool.Init(func() interface{} {
+		w := &fetchAttempt{session: p.session}
+		// NB(r): Bind attemptFn once to avoid creating receiver
+		// and function method pointer over and over again
+		w.attemptFn = w.perform
+		w.reset()
+		return w
+	})
+	p.initialized = true
```
Comment (xichen2020) [inline client/fetch_attempt.go]: This `initialized` flag seems to be only assigned to but never used?

File: client/options.go [hunk 2]
```diff
@@ -74,7 +75,7 @@ const (
 	defaultIdentifierPoolSize = 8192
 
 	// defaultWriteOpPoolSize is the default write op pool size
-	defaultWriteOpPoolSize = 1000000
+	defaultWriteOpPoolSize = 262144
```
Comment (xichen2020) [inline client/options.go]: why reducing the size to 256K out of curiosity?

File: client/fetch_attempt.go (hunk 1 above)
Comment (robskillington) [inline client/fetch_attempt.go]: I want to use this later to do deferred pool initialization, however the way the code is structured makes that difficult.  I can do this in a follow up PR that restructures a lot of the session.

File: client/options.go (hunk 2 above)
Comment (robskillington) [inline client/options.go]: After looking at how many writes are open at any given time (i.e. num goroutines) we seem to peak at 60-80k, the original number was arbitrarily picked.

File: client/fetch_attempt.go (hunk 1 above)
Comment (xichen2020) [inline client/fetch_attempt.go]: Yeah in that case would you mind removing the flag now and add it back in the followup PR then?

File: client/options.go (hunk 2 above)
Comment (xichen2020) [inline client/options.go]: Makes sense

**Pull request merged**